pysmt/smtlib/parser.py

```diff
@@ -123,153 +125,159 @@ def unbind_all(self, values):
             self.unbind(k)
 
 
-class Tokenizer(object):
-    """
-    Takes a file-like object and produces a stream of tokens following
+
+def tokenizer(handle, interactive=False):
+    """Takes a file-like object and produces a stream of tokens following
     the LISP rules.
-    """
 
-    def __init__(self, handle):
-        self.handle = handle
-        self.spaces = set([" ", "\n", "\t"])
-        self.separators = set(["(", ")", "|"])
-        self.specials = self.spaces | self.separators | set([";", ""])
-        self.read = self.handle.read
-        self.eof = False
-
-    def tokens(self):
-        c = self.read(1)
-        while not self.eof:
-            if c in self.specials:
-                # consume all the spaces
-                if c in self.spaces:
-                    while c and c in self.spaces:
-                        c = self.read(1)
-
-                elif c in self.separators:
-                    if c == "|":
-                        s = []
-                        c = self.read(1)
-                        while c and c != "|":
-                            s.append(c)
-                            c = self.read(1)
-                        if not c:
-                            raise SyntaxError("Expected '|'")
-                        yield ("".join(s))
-                    else:
-                        yield c
-                    c = self.read(1)
+    If interative is True, the file reading proceeds char-by-char with
+    no buffering. This is useful for interactive use for example with
+    a SMT-Lib2-compliant solver
+    """
+    spaces = set([" ", "\n", "\t"])
+    separators = set(["(", ")", "|"])
+    specials = spaces | separators | set([";", ""])
+
+    if not interactive:
+        reader = itertools.chain.from_iterable(handle) # reads char-by-char
+    else:
+        reader = interactive_char_iterator(handle)
+    c = next(reader)
+
+    eof = False
+    while not eof:
+        if c in specials:
+            # consume all the spaces
+            if c in spaces:
+                while c and c in spaces:
```

Comment:
This loop can be moved at the beginning of the outer ```while```, right?

```
while not eof:
    while c and c in spaces:
         c = next(reader)
```

If this is the case, we can probably replace this with an ```ifilter```